GRAPHICS: Speed up and refactor SVG handling

mikrosk opened the pull request:
Maybe you are aware of other places in code like this?

(I found this one the hard way)

lephilousophe reviewed (approved):
LGTM.
I am not sure other cases would be easy to find.
The policy has always been to use Surface::free after Surface::create and not to use Surface::free for other cases.

sev- commented:
I am confused by this change. Yes, `tmp.free()` is redundant. But what are those `_cache`-related changes about?

mikrosk commented:
@sev- `tmp` is initialised with `_cache` (which is allocated a few lines above): https://github.com/scummvm/scummvm/pull/4773/files#diff-42f6acbe42c71148cf49bffeec7be5d8572dd200e36f539149cf4a5398c60ccbL91

sev- commented:
Well, I understand that, but shouldn't it be split in two commits then?

mikrosk commented:
Now _I_ am confused. :) `tmp.free()` isn't redundant, `Surface`'s d-tor doesn't call `::free()` automatically (it has a default d-tor). Perhaps you are confusing it with a `ManagedSurface` ?

So if I'm guessing right, `tmp.free()` was supposed to release the `init`ialised `_cache` buffer, therefore this commit does that, the right way

mikrosk commented:
On a second look... perhaps there is something to it. That `_cache` isn't just a temporary buffer, it is supposed to hold the rasterized image of size `dw` * `dh` and if the size is the same, there is no reason to dealloc and alloc again and again.

Good catch!

mikrosk commented:
@sev- it seems there even was a memory leak, please take a look at the latest version whether the fix is good to go now.

lephilousophe commented:
I would like to take time to look at it on a proper screen.
Thanks

lephilousophe commented:
In fact we are leaking memory and I don't see why we need _cache, _cachedX, _cachedY at all.
I have a commit on lephilousophe/scummvm@ec2f1a96c773b392792c34a833ef66db59aa05fc which seems to work correctly and avoids the intermediate blitting.

mikrosk commented:
@lephilousophe isn't possible that the intermediate surface has been used for implicit conversion from, say, 16bpp to 32bpp (which the nvg rasterizer expects) ?

lephilousophe commented:
I thought that at first but both surfaces (`tmp` which uses `_cache` and `_render`) are created with `_pixelformat` which is always 32bpp as set in constructor and is never changed (no setter).
But I ask for confirmation because I don't know well this part of the code, maybe I missed something.

mikrosk commented:
Encouraged by @lephilousophe's analysis I have taken a further step, removed all the shenanigans with allocating and blitting, code is much faster and cleaner now.

As a bonus, I have also reverted one commit which doesn't seem to be necessary, so we are caching bitmaps during initial startup again.

All together this brought a speedup of 8 seconds on an intentionally weak hardware so I think it's worth applying. Please take a look whether I didn't miss something obvious.

lephilousophe reviewed (commented):
So, all in all, before this change we had a `SVGBitmap` class which would render on any surface size with a cache for the last rendered resolution.
This implied that we keep a decoder around and a rasterized cache.
This code was used in 2 places: theme engine and grid widget where the SVGBitmap was rendered once and destroyed. A texture of the same pixel format was used to store the data.

With new code, the SVG is not a renderer anymore but directly a surface with a size set at creation time and a fixed pixel format.
This is OK for current uses but may reduce hypothetical future use cases.
The SVGSurface now consists only in a ManagedSurface so the memory footprint is reduced.
There are now less blitting too.

Except from my inline comments, code looks good to me and the features removal seem acceptable to me.

mikrosk commented on gui/ThemeEngine.cpp:
```diff
@@ -650,11 +650,7 @@ bool ThemeEngine::addBitmap(const Common::String &filename, const Common::String
-		surf = nullptr;
-
-		_bitmaps.erase(filename);
+		return true;
```
Yes, I have been worrying about the same thing. However I have also been unable to find a use case where this causes trouble (I was switching themes via the GUI, maybe there is some other way?). @sev- if you can remember a scenario which was broken without this change, that would help.

lephilousophe replied:
You should switch the scale factor, not the theme.

mikrosk replied:
That, too - at least I couldn't find anything else than the large/medium/small scale options. And I don't see any difference in behavior.

lephilousophe replied:
OK. I see the why of the commit.
It was done in April 2021.
At that time (before ad31dfc8d5e35e7449587ccca4d81835beb16872 in August 2021), the _scaleFactor was depending on the height of the window in `gui/gui-manager.cpp`.

A resize of the window implied a `_scaleFactor` change which would have resulted in garbled images.
Changing the scale factor in GUI forces a `ThemeEngine` destroy and recreate, that's why we can't replicate the issue.

I still think we should take the safe side and clean the cache in `setBaseResolution`: changing the `_scaleFactor` has to invalidate the cache even if today we don't see the bug.

mikrosk commented on graphics/svg_surface.cpp:
```diff
@@ -18,76 +18,50 @@
-
 #ifdef SCUMM_BIG_ENDIAN
-	_pixelformat = new Graphics::PixelFormat(4, 8, 8, 8, 8, 24, 16, 8, 0);
+const Graphics::PixelFormat PIXELFORMAT = Graphics::PixelFormat(4, 8, 8, 8, 8, 24, 16, 8, 0);
```
Right. I've changed it in a slightly different way, using a `#define` as it serves basically the same purpose.

Otherwise I can inline the `#ifdef` in the constructor but it wont look as good.

sev- commented on graphics/svg_surface.cpp:
```diff
@@ -0,0 +1,67 @@
+	in->read(data, size);
+	data[size] = '\0';
+
+	NSVGimage *svg = nsvgParse(data, "px", 96);
```
Where does the cache go? Why? It is used extensively by our GUI code, and re-rendering it every time is not optimal, particularly in the icons list.

Rendering SVG is very expensive because it uses double-precision FPU arithmetics.

mikrosk replied:
Indeed. However from what I could see from the profiling, this bitmap is not reused. So while it looks good on the first sight, in reality this cache didn't provide any speedup, on the contrary.

As written in the other comment, I physically measured the speed this brought and it was enormous.

mikrosk replied:
I guess your discomfort with this change is related to the debate I had with @lephilousophe earlier -- whether it's acceptable to narrow SVGBitmap's features to its current usage (load svg file, render, copy and trash) or to keep it as generic as possible.

As the former approach brought significant speedup and simplified other parts of ScummVM code, too it looked like a good idea to go ahead.

sev- reviewed (commented):
I see no reasoning provided, nor what exactly does this speed up. Please, explain.

> This reverts commit https://github.com/mikrosk/scummvm/commit/04f040afce37a57948df6bf464ca8cb2453d0c22 which claims to fix a theme reload bug but
> I wasn't able to reproduce this behavior at all (with changed scale
> and/or renderer ThemeEngine is reset and its _bitmap cache doesn't
> contain any leftovers).

I do not understand why you express this discontent edging aggression in this commit log message. "I was not able to reproduce it, the commit log is lying and the commit is useless. Reverting it"

mikrosk commented:
> I do not understand why you express this discontent edging aggression in this commit log message. I read it as: "I was not able to reproduce it, the commit log is lying and the commit is useless. Reverting it"

That's definitely not the way I intended to be understood. It's only what I wrote there - this was fixing something but that something seems to be gone, bitmap loading is working (and cached) without this commit, period. As you can see, I tried to verify it with others to be sure that this claim is correct.

No need to read between the lines. :)

mikrosk commented on graphics/module.mk:
```diff
@@ -48,7 +48,7 @@ MODULE_OBJS := \
 	sjis.o \
 	surface.o \
-	svg.o \
+	svg_surface.o \
```
Just to follow the convention, so SVGSurface is located in svg_surface.

mikrosk commented on graphics/svg_surface.cpp:
```diff
@@ -0,0 +1,67 @@
+#define PIXELFORMAT Graphics::PixelFormat(4, 8, 8, 8, 8, 0, 8, 16, 24)
+#endif
+
+Graphics::SVGSurface::SVGSurface(Common::SeekableReadStream *in, int dw, int dh)
```
Well, it's inherited from ManagedSurface now, would you prefer to keep its old name despite this change?

sev- replied:
Yes, I prefer to keep the old name for consistency.

Why is it now inherited from ManagedSurface and not a plain Surface?

mikrosk replied:
Both ThemeEngine's `_bitmaps[filename]` and grid's `loadSurfaceFromFile` work with ManagedSurface so it was a nice optimization -- we can directly store/return `new`ly created instances.

mikrosk commented:
After talking to @sev-:

- rebased
- agreed that getting rid of `_cache` is good so I kept @lephilousophe's change
- agreed to keep the original name `SVGBitmap` for easier review
- rephrased the offending commit's language to more neutral tone + added @lephilousophe's observation

sev- commented:
Thanks!
